File: microservices/api/api-scala/app/ApplicationLoader.scala
import play.api._
import play.api.ApplicationLoader.Context
import play.api.routing.Router
import play.filters.HttpFiltersComponents
import play.filters.cors.CORSComponents
import play.api.mvc._
import play.api.libs.ws.ahc.AhcWSComponents
import modules.{AuthModule, SwaggerModule}
import router.Routes

/**
 * Application loader with compile-time dependency injection
 */
class ApplicationLoader extends play.api.ApplicationLoader {
  def load(context: Context): Application = {
    LoggerConfigurator(context.environment.classLoader).foreach {
      _.configure(context.environment, context.initialConfiguration, Map.empty)
    }
    new ApplicationComponents(context).application
  }
}

/**
 * Application components
 */
class ApplicationComponents(context: Context)
    extends BuiltInComponentsFromContext(context)
    with HttpFiltersComponents
    with CORSComponents
    with AhcWSComponents
    with _root_.controllers.AssetsComponents {

  // Load modules
  private val authModule = new AuthModule(environment, configuration)
  private val swaggerModule = new SwaggerModule(environment, configuration)
  
  // Services
  lazy val jwtService = new services.JwtService(configuration)(executionContext)
  lazy val authAction = new actions.AuthAction(playBodyParsers.default, jwtService)(executionContext)
  
  // Controllers
  lazy val healthController = new controllers.HealthController(controllerComponents, configuration)
  lazy val authController = new controllers.AuthController(controllerComponents, wsClient, configuration)(executionContext)
  lazy val userController = new controllers.UserController(controllerComponents, authAction, jwtService)(executionContext)
  lazy val dataController = new controllers.DataController(controllerComponents, authAction, wsClient, configuration)(executionContext)
  lazy val apiDocController = new controllers.ApiDocController(controllerComponents)
  
  // Routes
  lazy val router: Router = new Routes(
    httpErrorHandler,
    healthController,
    assets,
    apiDocController,
    authController,
    userController,
    dataController
  )
  
  // Override http filters
  override lazy val httpFilters: Seq[play.api.mvc.EssentialFilter] = {
    super.httpFilters ++ Seq(corsFilter)
  }
}

File: microservices/api/api-java/app/ApplicationLoader.java
import play.Application;
import play.ApplicationLoader;
import play.BuiltInComponentsFromContext;
import play.LoggerConfigurator;
import play.filters.components.HttpFiltersComponents;
import play.filters.cors.CORSComponents;
import play.routing.Router;
import router.Routes;

/**
 * Application loader with compile-time dependency injection
 */
public class ApplicationLoader implements play.ApplicationLoader {
    
    @Override
    public Application load(Context context) {
        LoggerConfigurator.apply(
            context.environment().classLoader()
        ).ifPresent(loggerConfigurator ->
            loggerConfigurator.configure(
                context.environment(), 
                context.initialConfig(), 
                java.util.Map.of()
            )
        );
        
        return new ApplicationComponents(context).application();
    }
    
    /**
     * Application components
     */
    static class ApplicationComponents extends BuiltInComponentsFromContext 
            implements HttpFiltersComponents, CORSComponents {
        
        public ApplicationComponents(Context context) {
            super(context);
        }
        
        // Services
        private final services.JwtService jwtService = 
            new services.JwtService(config());
            
        private final actions.AuthAction authAction = 
            new actions.AuthAction(jwtService);
        
        // Controllers
        private final controllers.HealthController healthController = 
            new controllers.HealthController(config());
            
        private final controllers.AuthController authController = 
            new controllers.AuthController(wsClient(), config());
            
        private final controllers.UserController userController = 
            new controllers.UserController(jwtService);
            
        private final controllers.DataController dataController = 
            new controllers.DataController(wsClient(), config());
            
        private final controllers.ApiDocController apiDocController = 
            new controllers.ApiDocController();
        
        @Override
        public Router router() {
            return new Routes(
                httpErrorHandler(),
                healthController,
                assets(),
                apiDocController,
                authController,
                userController,
                dataController
            );
        }
    }
}
